core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/AkkaClusterContainerResourceManager.scala

```diff
@@ -0,0 +1,376 @@
+
+  //cachedValues
+  var idMap: immutable.Set[Int] = Set.empty
+
+  def activationStartLogMessage(): String =
+    s"node stats ${clusterActionHostStats} reserved ${localReservations.size} (of max ${poolConfig.clusterManagedResourceMaxStarts}) containers ${reservedSize}MB " +
+      s"${reservedStartCount} pending starts ${reservedStopCount} pending stops " +
+      s"${scheduledStartCount} scheduled starts ${scheduledStopCount} scheduled stops"
+
+  def rescheduleLogMessage() = {
+    s"reservations: ${localReservations.size}"
+  }
+
+  def requestSpace(size: ByteSize) = {
+    val bufferedSize = size * 10 //request 10x the required space to allow for failures and additional traffic
+    //find idles up to this size
+    val removable = remoteUnused.map(u => u._2.map(u._1 -> _)).flatten.to[ListBuffer]
+    val removed = remove(removable, bufferedSize) //add some buffer so that any accumulating activations will have better chance
+    if (removed.nonEmpty) {
+      val removing = removed.groupBy(_._1).map(k => k._1 -> k._2.map(_._2)) //group the removables by invoker, will send single message per invoker
+      logging.info(this, s"requesting removal of ${removed.size} eligible idle containers ${removing}")
+      removing.foreach { r =>
+        clusterPoolData ! RequestReleaseFree(r._1, r._2)
+        //update unusedPool (we won't ask them to be removed twice)
+        remoteUnused = remoteUnused + (r._1 -> remoteUnused(r._1).filterNot(r._2.toSet))
+
+      }
+    }
+
+  }
+
+  def reservedSize = localReservations.values.map(_.size.toMB).sum
+  def remoteReservedSize = remoteReservations.values.map(_.map(_.size.toMB).sum).sum
+  def reservedStartCount = localReservations.values.count {
+    case p: Reservation => p.size.toMB >= 0
+    case _              => false
+  }
+  def reservedStopCount = localReservations.values.count {
+    case p: Reservation => p.size.toMB < 0
+    case _              => false
+  }
+  def scheduledStartCount = localReservations.values.count {
+    case p: Reservation => p.size.toMB >= 0
+    case _              => false
+  }
+  def scheduledStopCount = localReservations.values.count {
+    case p: Reservation => p.size.toMB < 0
+    case _              => false
+  }
+  private var clusterResourcesAvailable
+    : Boolean = false //track to log whenever there is a switch from cluster resources being available to not being available
+
+  def canLaunch(memory: ByteSize, poolMemory: Long, poolConfig: ContainerPoolConfig): Boolean = {
+
+    val localRes = localReservations.values.map(_.size) //active local reservations
+    val remoteRes = remoteReservations.values.toList.flatten.map(_.size) //remote/stale reservations
+
+    val allRes = localRes ++ remoteRes
+    //make sure there is at least one node with unreserved mem > memory
+    val canLaunch = clusterHasPotentialMemoryCapacity(memory.toMB, allRes) //consider all reservations blocking till they are removed during NodeStatsUpdate
```

Comment:
Not really, since this doesn't actually schedule containers to the cluster, it just estimates the potential ability of ContainerFactory to do so. If there are more resource types to consider in the future, this signature will need to change.

Reply:
> If there are more resource types to consider in the future

Reply:
> If there are more resource types to consider in the future

IIRC `disk` is something needed shortly ?

Reply:
This PR is not changing the scope of resource scheduling, it remains only based on memory for this PR.